h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -937,90 +1026,76 @@ private Chunk getChunkIfFound(long pos) {
      * <p>
-     * For in-memory stores, this method increments the version.
+     * This method does nothing if there are no unsaved changes,
+     * otherwise it increments the current version
+     * and stores the data (for file based stores).
      * <p>
-     * For persistent stores, it also writes changes to disk. It does nothing if
-     * there are no unsaved changes, and returns the old version. It is not
-     * necessary to call this method when auto-commit is enabled (the default
+     * It is not necessary to call this method when auto-commit is enabled (the default
      * setting), as in this case it is automatically called from time to time or
      * when enough changes have accumulated. However, it may still be called to
      * flush all changes to disk.
-     *
-     * @return the new version
-     */
-    public synchronized long commit() {
-        if (fileStore != null) {
-            return commitAndSave();
-        }
-        long v = ++currentVersion;
-        setWriteVersion(v);
-        return v;
-    }
-
-    /**
-     * Commit all changes and persist them to disk. This method does nothing if
-     * there are no unsaved changes, otherwise it increments the current version
-     * and stores the data (for file based stores).
      * <p>
      * At most one store operation may run at any time.
      *
      * @return the new version (incremented if there were changes)
      */
-    private synchronized long commitAndSave() {
-        if (closed) {
-            return currentVersion;
-        }
-        if (fileStore == null) {
-            throw DataUtils.newIllegalStateException(
-                    DataUtils.ERROR_WRITING_FAILED,
-                    "This is an in-memory store");
-        }
-        if (currentStoreVersion >= 0) {
-            // store is possibly called within store, if the meta map changed
-            return currentVersion;
-        }
-        if (!hasUnsavedChanges()) {
-            return currentVersion;
-        }
-        if (fileStore.isReadOnly()) {
-            throw DataUtils.newIllegalStateException(
-                    DataUtils.ERROR_WRITING_FAILED, "This store is read-only");
-        }
-        try {
-            currentStoreVersion = currentVersion;
-            currentStoreThread = Thread.currentThread();
-            return storeNow();
-        } finally {
-            // in any case reset the current store version,
-            // to allow closing the store
-            currentStoreVersion = -1;
-            currentStoreThread = null;
-        }
-    }
-
-    private long storeNow() {
-        try {
-            return storeNowTry();
-        } catch (IllegalStateException e) {
-            panic(e);
-            return -1;
+    public long commit() {
+        // unlike synchronization, this will also prevent re-entrance,
+        // which may be possible, if the meta map have changed
+        if (currentStoreThread.compareAndSet(null, Thread.currentThread())) {
```

Comment:
this seems tricky? if two different threads call commit() simultaneously, one of them is just going to return immediately, potentially before data has hit disk

Reply:
Yes, I do believe that "D" in ACID is a false promise, not worth paying a high price in most cases.
If another commit() is in progress, your data were most likely saved already. We all know that "hit the disk" in reality means "hit the disk's cache".

Reply:
yeah, but in this case we're going to return early, long before it's even hit the disk cache, which is a little dicky

Reply:
I hear you. So maybe we can make it an option, because someone may choose performance instead of a bigger "D"? It is similar to how often sync() should be called.

Reply:
I'd be fine with you creating a new method that works like this, and using it internally, and in H2, it's just that this is public API, and some people are using MVStore directly, and it seems like it might be a nasty surprise if they update and the semantics change like this?

Reply:
Or maybe we can just write it in big letters in the changelog, but there should at least a compatibility method people can use for a time if they want the old semantics.

Reply:
What about commit() always saves, and add a method commitConcurrently() (or so) with the new behavior? I think if people call commit() too often (from multiple threads) it's their problem, and not something we need to optimize for. But if you optimize for that case, I would do that in a new method.

But I'm not sure if this aspect is unrelated to this issue (it seems to me), so I would try to do it in a different issue.

Reply:
I agree that from pure MVStore perspective two methods is a clean solution, although commitConcurrently() is a bit confusing to me, commitIfNoneInProgress() is probably too verbose, tryCommit()?.
Lets now consider H2 as a client of MVStore. Should we use one or the other? It's up to the client (of H2 now), and it's probably should be a setting, switching between the two.

Reply:
Yes commitConcurrently is confusing... tryCommit is OK, if the Javadocs are clear. I guess H2 can usually call tryCommit, but maybe TransactionStore.close should call commit?